Bluetooth: services: Adds directory listing object to OTS

Thalley opened the pull request:
The directory listing object is an internal object which
content is the aggregation of all the metadata of all objects
(including the directory listing object itself). The client
can read this value to get a list of all objects with names,
lengths, and other metadata.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: services: Adds directory listing object to OTS

The directory listing object is an internal object which
content is the aggregation of all the metadata of all objects
(including the directory listing object itself). The client
can read this value to get a list of all objects with names,
lengths, and other metadata.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

jhedberg commented on subsys/bluetooth/services/ots/ots_dir_list.c:
```diff
@@ -0,0 +1,242 @@
+	net_buf_simple_add_le32(net_buf, obj->metadata.props);
+
+	/* Update the record length at the beginning */
+	sys_put_le16(net_buf->len - start_len, record_length_p);
```
Instead of using a `start_len` variable, it'd be cleaner to store the pointer returned by the `net_buf_simple_add(net_buf, sizeof(uint16_t))` call on line 38 since that's the same as `net_buf->len - start_len`.

jhedberg replied:
Sorry, I mean store the value to record_length_p from the return value of `net_buf_simple_add()`

jhedberg replied:
Also, I don't think we have the habit of encoding the variable type in the variable name, so I'd drop the `_p` from the end

jhedberg replied:
Another option would be to use `net_buf_simple_reserve(net_buf, sizeof(uint16_t))` and then you can use `net_buf_simple_push_le16(net_buf, net_buf->len)` without the need for `start_len` or `record_length_p` at all. I think that would be the cleanest, assuming there's nothing else in the buffer (i.e. `net_buf->len` will be the amount of data that you've added)

Thalley replied:
We need the `start_len` to calculate the length of the record once we are done inserting the values. 

We already store the pointer of when the object starts (`record_length_p`), but we don't have have the pointer to the tail (i.e. where the record ends), nor can I see any net_buf_simple API to get that. 

So I'm not sure how we can calculate the length of the record with pointers, instead of a start and end length.

jhedberg replied:
@Thalley see my comment that I posted almost simultaneously with yours. Does the buffer contain anything else in the beginning of this function?

jhedberg replied:
@Thalley I see this being called from a do-while loop, so I suppose the answer is "yes, the buffer may already contain other data"?

jhedberg replied:
@Thalley that said, I see you use the same "placeholder" pattern elsewhere. Please keep in mind the possibility of using `net_buf_simple_reserve()` + `net_buf_simple_push()` as an alternative pattern.

Thalley replied:
It may. This function is called for each object, so the content of the buffer is basically an encoded array of records, each with a length field (as the first field per record). 

So we have 
`[ {length, field, field, field}, {length, field, field, field}, ...] `
where each record `{ ... }` is added per function call. 

Instead of `start_len` we could do 
```
uint8_t *start = net_buf_simple_add(net_buf, sizeof(uint16_t));
...
...
sys_put_le16(net_buf_simple_tail(net_buf) - start, start);
```

Would that be better?

jhedberg replied:
@Thalley yeah, I think that's a little better since it explicitly associates `start` and the call to `net_buf_simple_add()`, which initialising from `net_buf->data` doesn't do (at least as clearly)

Thalley replied:
Agreed. Updated now :)

jhedberg commented on subsys/bluetooth/services/ots/ots_dir_list.c:
```diff
@@ -0,0 +1,239 @@
+	net_buf_simple_add_le32(net_buf, obj->metadata.props);
+
+	/* Update the record length at the beginning */
+	sys_put_le16(net_buf_simple_tail(net_buf) - start, start);
```
Nice, but now you're not using `start_len` anymore, i.e. you should probably remove it.

Btw, just double-checking: the length field is supposed to include the length of the length field itself and not only the data that follows it (that latter is more common in other Bluetooth protocols I'm familiar with)

Thalley replied:
Right :D 

Just double checked the spec: 
> Each Object Record shall begin with a “Length of Object Record” field that specifies the total size of the current Object Record in octets. The total value reported includes the 2 octets used by the “Length of Object Record” field itself.

So that should be right :)

kapi-no commented on subsys/bluetooth/services/ots/ots_dir_list.c:
```diff
@@ -0,0 +1,234 @@
+	sys_put_le16(net_buf_simple_tail(net_buf) - start, start);
+}
+
+void bt_ots_dir_list_add(struct bt_ots *ots,
```
Consider using obj_add/obj_remove suffix

I initially thought that this API creates the Directory Listing Object in the object pool.

Thalley replied:
Good idea

kapi-no replied:
I would still leave init() and selected() API functions with obj infix. I know that the Directory listing is also an object but I think its better to distinguish between API functions that operate on the external object (non Directory listing object) and API function that operate only on the Directory Listing object.

Thalley replied:
Alright; done.

kapi-no commented on subsys/bluetooth/services/ots/ots_internal.h:
```diff
@@ -66,6 +70,12 @@ struct bt_gatt_ots_indicate {
 	bool is_enabled;
 };
 
+struct bt_ots_dir_list {
+	struct net_buf_simple net_buf;
+	struct bt_gatt_ots_object *dir_list_obj;
+	uint8_t _content[DIR_LIST_MAX_SIZE];
+};
```
Please move this definition to ots_dir_list header and simply include it here (same for new defines).

Thalley replied:
That was my original intend as well, but then we'd have to deal with circular dependencies as this needs the declaration of struct bt_gatt_ots_object.

Possible solutions, if we still want to move it, is to add an opaque type or to use void for the obj pointer type.

kapi-no replied:
Hmm, thanks for explaining.

We could have ots_object_internal.h header with Object structure definition and all its deps.

However, we would still have a circular dependency because the OTS instance is used in the ots_dir_list_internal.h. Is it possible to refactor the ots_dir_list_internal API so that it does not have to access the full OTS instance? I think you are using only three instance fields in the Directory listing source file: dir_list, cur_obj and obj_manager.

I also think that it is a good practice to pass only data that are required by the receiving module and nothing more. I know that access to the full OTS state is convenient but it could also be more error-prone. For example, you may by accident change the OACP state in the Directory listing related code.

Thalley replied:
I am only using a few instance fields indeed, so it should be possible. 

On the other hand, if the function requires e.g. the `obj_manager`, the `dir_list` and `cur_obj`, there is likewise a risk that one of these parameters might be passed from the wrong OTS instance, and we would have to (or should) add checks/asserts that the different values are from the same OTS instance, which would basically do a reverse lookup (using e.g. `CONTAINER_OF`) to get the OTS instance. (in reality we could do that as a solution as well, and just take the dir_list object as arguments, but that's also kinda weird in my opinion).

kapi-no replied:
Yes, sure, I can see your point. I still think that it would be useful to change the code according to my suggestion.

Any opinions on this from the other reviewers? I think that outside feedback would help to resolve this discussion faster.

Thalley replied:
@kapi-no What do you think? Should I refactor to only require the needed parameters, or keep as is?

kapi-no replied:
@Thalley, I think it is better to refactor it as I laid out in my suggestion

Thalley replied:
@kapi-no I refactored it, and added the modifications in a new commit.

kapi-no reviewed (commented):
Looking good, a few comments from me

Thalley commented on subsys/bluetooth/services/ots/ots_dir_list.c:
```diff
@@ -0,0 +1,234 @@
+			if (!dir_lists[i].dir_list_obj) {
+				ots->dir_list = &dir_lists[i];
+			}
+		}
```
I'm adding an assert instead, as this would be an internal stack issue and not something the application can deal with.

Thalley commented on subsys/bluetooth/services/ots/ots.c:
```diff
@@ -210,11 +228,23 @@ int bt_ots_obj_delete(struct bt_ots *ots, uint64_t id)
 	}
 
+#if IS_ENABLED(CONFIG_BT_OTS_DIR_LIST_OBJ)
+	if (ots->dir_list->dir_list_obj->state.type !=
+			BT_GATT_OTS_OBJECT_IDLE_STATE) {
```
Found several other lines that can be merged. Thanks!

Thalley commented on subsys/bluetooth/services/ots/Kconfig:
```diff
@@ -13,14 +13,35 @@ config BT_OTS
 config BT_OTS_MAX_OBJ_CNT
-	int "Maximum number of objects that each service instance can store"
+	hex "Maximum number of objects that each service instance can store"
 	default 5
```
Fixed

Thalley commented on subsys/bluetooth/services/ots/Kconfig:
```diff
@@ -13,14 +13,35 @@ config BT_OTS
-	int "Maximum number of objects that each service instance can store"
+	hex "Maximum number of objects that each service instance can store"
 	default 5
+	range 0x2 0x17D if BT_OTS_DIR_LIST_OBJ
```
Fixed

kapi-no replied:
0x017D (byte aligned)

Thalley commented on subsys/bluetooth/services/ots/Kconfig:
```diff
@@ -13,14 +13,35 @@ config BT_OTS
 if BT_OTS
 
+config BT_OTS_DIR_LIST_OBJ
+	bool "Enables the directory listing object"
```
Fixed

Thalley commented on subsys/bluetooth/services/ots/Kconfig:
```diff
@@ -13,14 +13,35 @@ config BT_OTS
+	help
+	  Enables the directory listing object, which is an object that
+	  contains all the metadata from all other objects, for easy exposure
+	  to a client. Enabling this will use on of the objects given by
```
Fixed

Thalley commented on subsys/bluetooth/services/ots/Kconfig:
```diff
@@ -13,14 +13,35 @@ config BT_OTS
 	default 1
 	range 1 1 if !BT_OTS_SECONDARY_SVC
 
+# Given the maximum size of a directory listing item (172) and the maximum size
+# of an object using the net_buf implementation is 2^16-1, the maximum number
+# of objects is given by 2^16-1 / 172 = 381 == 0x17D)
```
Fixed

Thalley commented on subsys/bluetooth/services/ots/Kconfig:
```diff
@@ -13,14 +13,35 @@ config BT_OTS
+
+config BT_OTS_DIR_LIST_OBJ_NAME
+	string "The object name of the directory listing object"
+	default "Directory_listing"
```
Fixed

Thalley commented on subsys/bluetooth/services/ots/ots_dir_list.c:
```diff
@@ -0,0 +1,234 @@
+
+		err = bt_gatt_ots_obj_manager_next_obj_get(ots->obj_manager,
+							   obj, &obj);
+
```
Fixed

jhedberg commented on subsys/bluetooth/services/ots/ots_dir_list.c:
```diff
@@ -0,0 +1,227 @@
+	sys_put_le16(ots->dir_list->net_buf.len, ots->dir_list->net_buf.data);
+
+	/* Update Directory Listing Object metadata size */
+	ots->dir_list->dir_list_obj->metadata.size.cur =
+		ots->dir_list->net_buf.len;
```
Is the line split really necessary? On line 135 you have the exact same assignment on a single line.

Thalley replied:
Good catch - Missed that one :)

asbjornsabo commented on subsys/bluetooth/services/ots/ots_dir_list.c:
```diff
@@ -0,0 +1,226 @@
+		 "Could not assign Directory Listing Object to OTS instance %p", ots);
+
+	__ASSERT(strlen(dir_list_obj_name) <= UINT8_MAX,
+		 "BT_OTS_DIR_LIST_OBJ_NAME shall be less than 255 octets");
```
The assert says "less than or equal to" (<=), while the text says "less than".

asbjornsabo commented on subsys/bluetooth/services/ots/ots_dir_list.c:
```diff
@@ -0,0 +1,226 @@
+		return 0;
+	}
+
+	*data = ots->dir_list->net_buf.data;
```
What about the offset?
This returns the same pointer, no matter what offset parameter is given in.

Thalley replied:
Hah, yeah that's something important missing :D

asbjornsabo reviewed (changes requested):
(I have not completed my review, this is as far as I got for now.)

Thalley commented on include/bluetooth/services/ots.h:
```diff
@@ -193,6 +196,124 @@ enum {
+ *  @param flags Directory Listing Object flags.
+ */
+#define BT_OTS_DIR_LIST_GET_FLAG_TYPE_128(flags) \
+	((flags) & BT_OTS_DIR_LIST_FLAG_TYPE_128)
```
Nice catch!

Thalley commented on include/bluetooth/services/ots.h:
```diff
@@ -33,6 +33,9 @@ extern "C" {
 /** @brief Length of OTS object ID string (in bytes). */
 #define BT_OTS_OBJ_ID_STR_LEN 15
 
+/** @brief ID of the Directory Listing Object */
```
I think the general consensus is that we don't do that in the headers.

asbjornsabo replied:
OK

Thalley commented on include/bluetooth/services/ots.h:
```diff
@@ -193,6 +196,124 @@ enum {
 #define BT_OTS_OBJ_GET_PROP_MARKED(prop) \
 	((prop) & BIT(BT_OTS_OBJ_PROP_MARKED))
 
+/** @brief Directory Listing record flag field. */
```
See https://github.com/zephyrproject-rtos/zephyr/pull/32618#discussion_r584546914

Thalley commented on subsys/bluetooth/services/ots/ots_dir_list.c:
```diff
@@ -0,0 +1,226 @@
+	}
+
+	/* skip 16bits at the beginning of the record for the record's length */
+	start = net_buf_simple_add(net_buf, sizeof(uint16_t));
```
Not really: `net_buf_simple_reserve` will only reserve memory at the beginning of the buffer, and not where the tail is, so it cannot be used for here where we need to "reserve" the 16 bits for each record.

Thalley commented on subsys/bluetooth/services/ots/ots_dir_list.c:
```diff
@@ -0,0 +1,226 @@
+			 * the objects current location
+			 */
+			memmove(ots->dir_list->net_buf.data + offset + len,
+				ots->dir_list->net_buf.data + offset,
```
Oh, nice catch! I guess I must've though it was the typical "src, dst" ordering :| Unfortunately I don't have a client that can exercise all of these changes easily.

asbjornsabo replied:
My experience is limited, but quite a few have "dst, src" ordering: memmove, memcpy, memccpy, strcpy, 
So I thought that was the "typical" ordering.

I do believe I read in LWN (or similar) some years ago that someone had run a check on a large code base (the Linux kernel?), and there was a significant percentage of the cases which had the arguments swithced ...

Thalley commented on subsys/bluetooth/services/ots/ots_dir_list.c:
```diff
@@ -0,0 +1,226 @@
+	}
+
+	__ASSERT(offset <= ots->dir_list->net_buf.len, "Object was not removed");
+
```
I've added a function that validates whether an object is part of an object manager to check this.

Thalley commented on subsys/bluetooth/services/ots/ots_dir_list.c:
```diff
@@ -0,0 +1,226 @@
+	int err;
+
+	err = bt_gatt_ots_obj_manager_first_obj_get(ots->obj_manager, &obj);
+
```
Added `err` check in the assert.

Thalley committed:
Bluetooth: Services: Adds bt_gatt_ots_obj_manager_obj_contains

Adds a ots_obj_manager function that returns true if the object is
in the object manager, or false otherwise.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

asbjornsabo commented on subsys/bluetooth/services/ots/Kconfig:
```diff
@@ -13,14 +13,34 @@ config BT_OTS
-	default 5
+	hex "Maximum number of objects that each service instance can store"
+	default 0x05
+	# Given the maximum size of a directory listing item (172) and the maximum size of an object
```
Replace "item" with "record"? See comment further down.

Thalley replied:
Agreed

asbjornsabo reviewed (changes requested):
A couple of minor things more.

Thalley commented on subsys/bluetooth/services/ots/ots_internal.h:
```diff
@@ -16,6 +16,10 @@ extern "C" {
 #include "ots_oacp_internal.h"
 #include "ots_olcp_internal.h"
 
+/** Maximum size of the Directory Listing Object. Table 4.1 in the OTS spec. */
```
Changed to `Directory Listing Object Record`

asbjornsabo replied:
Does not seem to have changed.

Thalley replied:
Please refresh :D It's changed for me (and Github has also marked it as "Outdated")

Thalley committed:
Bluetooth: Services: Ensure correct ID before getting object index

Moved the call to obj_id_to_index such that we can check for valid
ID before getting the index.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented on subsys/bluetooth/services/ots/ots_obj_manager.c:
```diff
@@ -36,6 +33,32 @@ struct bt_gatt_ots_obj_manager {
 	bool is_assigned;
 };
 
+static uint64_t obj_id_to_index(uint64_t id)
```
That is a good point. There is a check, but it was after this call. The order here doesn't matter that much, but moved the call to after the check anyhow to optimize it a bit.

asbjornsabo reviewed (approved):
No further comments from me.

jhedberg commented on subsys/bluetooth/services/ots/Kconfig:
```diff
@@ -13,14 +13,34 @@ config BT_OTS
 config BT_OTS_MAX_OBJ_CNT
-	int "Maximum number of objects that each service instance can store"
-	default 5
+	hex "Maximum number of objects that each service instance can store"
```
Why `hex` instead of `int`? Usually it's more natural to represent "a number of something" in decimal rather than hexadecimal.

Thalley replied:
I agree. I changed it to make the ranges easier to compare to the specification. 

Depending on the BT_OTS_DIR_LIST_OBJ, there are 2 max ranges, and writing `0xFFFFFFFFFFFFFEFF` in decimal seemed wrong to me, but I can change it back again.

carlescufi commented:
@kapi-no could you take another look?

Thalley commented:
> @kapi-no could you take another look?

@carlescufi Based on the last comment from @kapi-no I think he/we wanted to hear input from other reviewers regarding the solution to the circular dependency.

kapi-no commented on include/bluetooth/services/ots.h:
```diff
@@ -193,6 +196,124 @@ enum {
+ *
+ *  @param flags Directory Listing Object flags.
+ */
+#define BT_OTS_DIR_LIST_GET_FLAG_CUR_SIZE(flags) \
+	((flags) & BIT(BT_OTS_DIR_LIST_FLAG_CUR_SIZE))
+
+/** @brief Get @ref BT_OTS_DIR_LIST_GET_FLAG_ALLOC_SIZE flag.
+ *
+ *  @param flags Directory Listing Object flags.
+ */
+#define BT_OTS_DIR_LIST_GET_FLAG_ALLOC_SIZE(flags) \
+	((flags) & BIT(BT_OTS_DIR_LIST_FLAG_ALLOC_SIZE))
+
+/** @brief Get @ref BT_OTS_DIR_LIST_GET_FLAG_FIRST_CREATED flag.
+ *
+ *  @param flags Directory Listing Object flags.
+ */
+#define BT_OTS_DIR_LIST_GET_FLAG_FIRST_CREATED(flags) \
+	((flags) & BIT(BT_OTS_DIR_LIST_FLAG_FIRST_CREATED))
+
+/** @brief Get @ref BT_OTS_DIR_LIST_GET_FLAG_LAST_MODIFIED flag.
+ *
+ *  @param flags Directory Listing Object flags.
+ */
+#define BT_OTS_DIR_LIST_GET_FLAG_LAST_MODIFIED(flags) \
+	((flags) & BIT(BT_OTS_DIR_LIST_FLAG_LAST_MODIFIED))
+
+/** @brief Get @ref BT_OTS_DIR_LIST_GET_FLAG_PROPERTIES flag.
+ *
+ *  @param flags Directory Listing Object flags.
+ */
+#define BT_OTS_DIR_LIST_GET_FLAG_PROPERTIES(flags) \
+	((flags) & BIT(BT_OTS_DIR_LIST_FLAG_PROPERTIES))
+
+/** @brief Get @ref BT_OTS_DIR_LIST_GET_FLAG_EXTENDED flag.
+ *
+ *  @param flags Directory Listing Object flags.
+ */
+#define BT_OTS_DIR_LIST_GET_FLAG_EXTENDED(flags) \
+	((flags) & BIT(BT_OTS_DIR_LIST_FLAG_EXTENDED))
```
I don't see any API to configure the Directory Listing object flags. The Directory listing object is created internally based on configuration. Can we move these macros and enums to the ots_dir_list_internal header?

Thalley replied:
Agreed

kapi-no commented on include/bluetooth/services/ots.h:
```diff
@@ -33,6 +33,9 @@ extern "C" {
 #define BT_OTS_OBJ_ID_STR_LEN 15
 
+/** @brief ID of the Directory Listing Object */
+#define OTS_OBJ_ID_DIR_LIST     0x000000000000
```
Why do we need to expose this information to the user? I think we can move this definition to the internal header as well

Thalley replied:
I guess the idea was that the object ID would be used for a potential client, but we can revisit that once we decide to add that. Moved to internal for now.

carlescufi commented:
@Thalley can you update this PR?

Thalley committed:
Bluetooth: services: Refacter OTS dir list to avoid circular dependency

Refactors the OTS directory listing object implementation to avoid
circular dependency.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

kapi-no reviewed (approved):
Thanks, LGTM
